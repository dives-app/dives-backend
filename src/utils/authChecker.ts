import { AuthChecker } from "type-graphql";
import { Context } from "../../types";
import { ApolloError } from "apollo-server-errors";

/**
 * Checks if user is authorized
 * @param userId userId passed by the Context
 * @param roles array of required roles to authorize
 * @throws if not authorized throws ApolloError - INVALID TOKEN
 * @returns true
 */
export const authChecker: AuthChecker<Context> = ({ context: { userId } }, roles) => {
  if (roles.length === 0) {
    if (!userId) throw new ApolloError("No user logged in", "INVALID_TOKEN");
    return true;
  }
  return true;
};
