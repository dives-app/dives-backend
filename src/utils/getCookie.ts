import { APIGatewayProxyEvent } from "aws-lambda";

/**
 * Searches for a cookie in APIGatewayProxyEvent and returns its value
 * @param event APIGatewayProxyEvent
 * @param cookieName cookie to search alphanumeric value + !#%&'-_`~
 * @returns cookie value or undefined if not found
 */
export function getCookie(event: APIGatewayProxyEvent, cookieName: string): string | undefined {
  const matcher = new RegExp(`${cookieName}=.*?(?=;|$)`, "m");
  return event.headers.Cookie?.match(matcher)?.[0]?.slice(cookieName.length + 1);
}
