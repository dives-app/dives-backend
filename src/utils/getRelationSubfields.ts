import { SelectionSetNode } from "graphql";
import { FieldNode } from "graphql/language/ast";

export function getRelationSubfields(
  selectionSet?: SelectionSetNode,
  parentNamespace: string = ""
): Array<string> {
  if (selectionSet) {
    return selectionSet.selections
      .filter(
        (node) =>
          "selectionSet" in node &&
          "name" in node &&
          node.selectionSet !== undefined
      )
      .flatMap((node: FieldNode) => {
        const currentFieldName = parentNamespace + node.name.value;
        return [
          currentFieldName,
          ...getRelationSubfields(node.selectionSet, `${currentFieldName}.`),
        ];
      });
  }
  return [];
}
