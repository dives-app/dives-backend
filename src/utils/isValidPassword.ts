/**
 * Validates a password to contain at least one digit, small letter, capital letter, special character and be at least 8 characters long
 * @param password password to validate
 * @returns true if password is valid, false otherwise
 */
export function isValidPassword(password: string) {
  // Contains at least one digit (\d)
  // Contains at least one small letter ([a-z])
  // Contains at least one capital letter ([A-Z])
  // Contains at least one special character (\W)
  // Has at least length of 8 characters
  const passwordRegExp = /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*\W).{8,}/;
  return passwordRegExp.test(password);
}
