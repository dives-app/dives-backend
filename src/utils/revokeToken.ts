import { Connection } from "typeorm";
import { User } from "../entities/User";

/**
 * Generates a new token invalidating the old one
 * @param userId ID of user to generate new token for
 * @param connection database connection
 */
export async function revokeToken(userId: string, connection: Connection) {
  await connection
    .createEntityManager()
    .update(User, { id: userId }, { token: () => "uuid_generate_v4()" });
}
