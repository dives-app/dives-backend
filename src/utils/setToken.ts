import { Cookie } from "../../types";

/**
 * Sends a token cookie with the next request
 * @param token token to be set
 * @param setCookies array of cookies to be set
 */
export async function setToken(token: string, setCookies: Array<Cookie>) {
  const { STAGE, SERVER_HOST } = process.env;
  const MILLISECONDS_IN_A_WEEK = 60_480_000;

  setCookies.push({
    name: "token",
    value: token,
    options: {
      domain: SERVER_HOST,
      expires: new Date(Date.now() + MILLISECONDS_IN_A_WEEK),
      httpOnly: true,
      maxAge: MILLISECONDS_IN_A_WEEK / 100,
      path: "/",
      sameSite: true,
      secure: STAGE !== "local",
    },
  });
}
