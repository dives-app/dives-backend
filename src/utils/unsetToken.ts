import { Cookie } from "../../types";

/**
 * Sends empty token cookie expiring immediately with the next request
 * @param setCookies array of cookies to be set
 */
export async function unsetToken(setCookies: Array<Cookie>) {
  const { STAGE, SERVER_HOST } = process.env;

  setCookies.push({
    name: "token",
    value: "",
    options: {
      domain: SERVER_HOST,
      httpOnly: true,
      // Set and max-age to unset the cookie in the browser
      maxAge: 0,
      path: "/",
      // Disable sameSite to allow localhost development with staging backend
      sameSite: STAGE === "staging" ? "none" : "strict",
      secure: STAGE !== "dev",
    },
  });
}
