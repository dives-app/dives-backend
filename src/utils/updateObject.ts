/**
 * Mutably updates the object using updateData
 * @param object object to mutably update
 * @param updateData part of object
 * @returns mutated object
 */
export function updateObject<T>(object: T, updateData: Partial<T>): T {
  Object.entries(updateData).forEach(([key, value]) => {
    if (value === undefined) return;
    object[key] = value;
  });
  return object;
}
