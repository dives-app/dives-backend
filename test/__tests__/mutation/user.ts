import { createTestClient } from "apollo-server-testing";
import { gql } from "apollo-server-lambda";
import { testServer } from "../../utils/testServer";
import { createTable, deleteTable } from "../../utils/setupAndTeardown";

beforeEach((done) => {
  createTable(done);
});
afterEach((done) => {
  deleteTable(done);
});

it("Creates user", async () => {
  expect.assertions(1);
  const CREATE_USER = gql`
    mutation {
      createUser(
        email: "test@user.com"
        password: "legitP@55"
        name: "Test User"
        birthDate: "2001-02-03"
        country: "PL"
      ) {
        id
        email
      }
    }
  `;
  const { mutate } = createTestClient(testServer);
  const res = await mutate({ mutation: CREATE_USER });
  expect(res.data.createUser.email).toBe("test@user.com");
});

it("Throws error on email collision", async () => {
  expect.assertions(1);
  const CREATE_USER = gql`
    mutation {
      createUser(
        email: "test@user.com"
        password: "legitP@55"
        name: "Test User"
        birthDate: "2001-02-03"
        country: "PL"
      ) {
        id
        email
      }
    }
  `;
  const { mutate } = createTestClient(testServer);
  await mutate({ mutation: CREATE_USER });
  const collisionResponse = await mutate({ mutation: CREATE_USER });
  expect(collisionResponse).toMatchSnapshot();
});
