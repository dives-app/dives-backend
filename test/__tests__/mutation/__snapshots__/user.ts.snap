// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Throws error on email collision 1`] = `
Object {
  "data": Object {
    "createUser": null,
  },
  "errors": Array [
    [GraphQLError: User with that email already exists],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
